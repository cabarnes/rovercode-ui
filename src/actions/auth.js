// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
