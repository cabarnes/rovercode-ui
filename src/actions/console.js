// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
