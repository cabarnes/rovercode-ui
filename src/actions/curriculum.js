// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
