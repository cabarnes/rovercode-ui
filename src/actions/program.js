// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
