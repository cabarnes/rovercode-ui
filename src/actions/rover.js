// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SCAN = 'SCAN';
export const SCAN_FULFILLED = `${SCAN}_FULFILLED`;
export const SCAN_REJECTED = `${SCAN}_REJECTED`;
export const CONNECT_ROVER = 'CONNECT_ROVER';
export const CONNECT_ROVER_FULFILLED = `${CONNECT_ROVER}_FULFILLED`;
export const CONNECT_ROVER_REJECTED = `${CONNECT_ROVER}_REJECTED`;
export const SEND_ROVER = 'SEND_ROVER';
export const SEND_ROVER_FULFILLED = `${SEND_ROVER}_FULFILLED`;
export const SEND_ROVER_REJECTED = `${SEND_ROVER}_REJECTED`;
export const DISCONNECT_ROVER = 'DISCONNECT_ROVER';

const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
const UART_TX_CHARACTERISTIC_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
const UART_RX_CHARACTERISTIC_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

// action creators
export const scan = () => ({
  type: SCAN,
  payload: navigator.bluetooth.requestDevice({
    filters: [{ namePrefix: 'BBC micro:bit' }],
    optionalServices: [UART_SERVICE_UUID],
  }),
});

export const connect = (device, onMessage) => ({
  type: CONNECT_ROVER,
  payload: device.gatt.connect()
    .then(server => server.getPrimaryService(UART_SERVICE_UUID))
    .then(service => Promise.all([
      service.getCharacteristic(UART_RX_CHARACTERISTIC_UUID),
      service.getCharacteristic(UART_TX_CHARACTERISTIC_UUID)
        .then((characteristic) => {
          characteristic.startNotifications();
          characteristic.addEventListener(
            'characteristicvaluechanged',
            onMessage,
          );
          return characteristic;
        }),
    ])),
});

export const send = (channel, message) => ({
  type: SEND_ROVER,
  payload: channel.writeValue(message),
});

export const disconnect = device => ({
  type: DISCONNECT_ROVER,
  payload: device.gatt.disconnect(),
});
