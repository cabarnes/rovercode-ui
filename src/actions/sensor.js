// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';
export const CHANGE_LIGHT_SENSOR_READINGS = 'CHANGE_LIGHT_SENSOR_READINGS';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});

export const changeLightSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LIGHT_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});
