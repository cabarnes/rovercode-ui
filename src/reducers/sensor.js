import {
  CHANGE_LEFT_SENSOR_STATE,
  CHANGE_RIGHT_SENSOR_STATE,
  CHANGE_LIGHT_SENSOR_READINGS,
  NOT_COVERED,
} from '../actions/sensor';

export default function sensor(
  state = {
    left: NOT_COVERED,
    right: NOT_COVERED,
    leftLightSensorReading: -1,
    rightLightSensorReading: -1,
  },
  action,
) {
  switch (action.type) {
    case CHANGE_LEFT_SENSOR_STATE:
      return {
        ...state,
        left: action.payload,
      };
    case CHANGE_RIGHT_SENSOR_STATE:
      return {
        ...state,
        right: action.payload,
      };
    case CHANGE_LIGHT_SENSOR_READINGS:
      return {
        ...state,
        leftLightSensorReading: action.payload.leftReading,
        rightLightSensorReading: action.payload.rightReading,
      };
    default:
      return state;
  }
}
