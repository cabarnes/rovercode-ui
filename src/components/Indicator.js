import React from 'react';
import { Grid } from '@material-ui/core';
import { styled, withStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import { COVERED } from '@/actions/sensor';

const mapStateToProps = ({ sensor }) => ({ sensor });

const Circle = styled('div')({
  margin: '0 auto',
  borderRadius: '25px',
  border: '1px solid #9A9A9A',
  width: '24px',
  height: '24px',
  opacity: 1,
});

const CircleCovered = withStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.success.main,
  },
}))(Circle);

const CircleNotCovered = withStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.background.default,
  },
}))(Circle);

const Indicator = ({ sensor }) => (
  <Grid container direction="row" justify="space-evenly">
    {
      /* eslint-disable react/no-array-index-key */
      [sensor.left, sensor.right].map((s, i) => (
        <Grid item key={i}>
          {
            s === COVERED ? (
              <CircleCovered id={`covered-${i}`} />
            ) : (
              <CircleNotCovered id={`not-covered-${i}`} />
            )
          }
        </Grid>
      ))
    }
  </Grid>
);

Indicator.propTypes = {
  sensor: PropTypes.shape({
    left: PropTypes.number.isRequired,
    right: PropTypes.number.isRequired,
  }).isRequired,
};

export default hot(module)(connect(mapStateToProps)(Indicator));
