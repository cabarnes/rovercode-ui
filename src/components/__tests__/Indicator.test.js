import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import { COVERED, NOT_COVERED } from '@/actions/sensor';
import Indicator from '../Indicator';

describe('The Indicator component', () => {
  const mockStore = configureStore();
  let store;

  test('renders on the page with no errors', () => {
    store = mockStore({
      sensor: {
        left: COVERED,
        right: NOT_COVERED,
      },
    });
    const wrapper = mount(<Indicator store={store} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays left covered', () => {
    store = mockStore({
      sensor: {
        left: COVERED,
        right: NOT_COVERED,
      },
    });
    const wrapper = shallow(<Indicator store={store} />).dive().dive();

    expect(wrapper.find('#covered-0').exists()).toBe(true);
    expect(wrapper.find('#not-covered-1').exists()).toBe(true);
  });

  test('displays right covered', () => {
    store = mockStore({
      sensor: {
        left: NOT_COVERED,
        right: COVERED,
      },
    });
    const wrapper = shallow(<Indicator store={store} />).dive().dive();

    expect(wrapper.find('#not-covered-0').exists()).toBe(true);
    expect(wrapper.find('#covered-1').exists()).toBe(true);
  });

  test('displays both covered', () => {
    store = mockStore({
      sensor: {
        left: COVERED,
        right: COVERED,
      },
    });
    const wrapper = shallow(<Indicator store={store} />).dive().dive();

    expect(wrapper.find('#covered-0').exists()).toBe(true);
    expect(wrapper.find('#covered-1').exists()).toBe(true);
  });

  test('displays neither covered', () => {
    store = mockStore({
      sensor: {
        left: NOT_COVERED,
        right: NOT_COVERED,
      },
    });
    const wrapper = shallow(<Indicator store={store} />).dive().dive();

    expect(wrapper.find('#not-covered-0').exists()).toBe(true);
    expect(wrapper.find('#not-covered-1').exists()).toBe(true);
  });
});
