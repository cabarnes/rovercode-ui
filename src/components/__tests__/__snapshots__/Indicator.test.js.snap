// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Indicator component renders on the page with no errors 1`] = `
<Connect(Indicator)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Indicator
    dispatch={[MockFunction]}
    sensor={
      Object {
        "left": 2,
        "right": 2,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Grid
      centered={true}
      columns={12}
    >
      <div
        className="ui centered twelve column grid"
      >
        <GridRow>
          <div
            className="row"
          >
            <GridColumn
              width={6}
            >
              <div
                className="six wide column"
              >
                <div
                  className="indicator not-covered"
                  id="leftIndicator"
                />
              </div>
            </GridColumn>
            <GridColumn
              width={6}
            >
              <div
                className="six wide column"
              >
                <div
                  className="indicator not-covered"
                  id="rightIndicator"
                />
              </div>
            </GridColumn>
          </div>
        </GridRow>
      </div>
    </Grid>
  </Indicator>
</Connect(Indicator)>
`;
